### PR TITLE
feat: allow custom v8 snapshots to be used in the main process and the default snapshot in the renderer process

### LINKED ISSUE #35170: [Feature Request]: Allow custom v8 snapshots to be used in the main process and the default snapshot in the browser process
### Preflight Checklist

- [X] I have read the [Contributing Guidelines](https://github.com/electron/electron/blob/main/CONTRIBUTING.md) for this project.
- [X] I agree to follow the [Code of Conduct](https://github.com/electron/electron/blob/main/CODE_OF_CONDUCT.md) that this project adheres to.
- [X] I have searched the [issue tracker](https://www.github.com/electron/electron/issues) for a feature request that matches the one I want to file, without success.

### Problem Description

We are using `electron-mksnapshot` to generate a custom v8 snapshot for our application that is desired to run **only in the main process**, not the renderer process. While debugging our app, we noticed that the snapshot is available to the renderer process **even with nodeIntegration disabled**. 

We noticed that electron is [patching](https://github.com/electron/electron/blob/main/patches/chromium/load_v8_snapshot_in_browser_process.patch?rgh-link-date=2022-08-02T00%3A24%3A21Z) chromium to always load the v8 snapshots, with a comment indicating that this is required. Is loading the v8 snapshot required even when nodeIntegration is disabled?

We would like for there to be a way to use our custom snapshot in the main process but use the default snapshot (or no snapshot with nodeIntegration disabled) in the browser process.

### Proposed Solution

This could be done a variety of ways, but one possibility would be to update the [patch](https://github.com/electron/electron/blob/main/patches/chromium/load_v8_snapshot_in_browser_process.patch?rgh-link-date=2022-08-02T00%3A24%3A21Z) to not load the custom snapshot when node integration is disabled. Or if it's desired sometimes to have node integration disabled **and** a v8 snapshot included, you could have an additional parameter that specifies as such.

### Alternatives Considered

One alternative for us would be to have separate main electron processes: one for the bulk of our main code that uses the snapshot, and one to purely launch the browser that doesn't launch the snapshot. This is undesirable as it significantly complicates the code and potentially bloats the size of our binary.

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by ryanthemanuel**
#### Description of Change

We are using electron-mksnapshot to generate a custom v8 snapshot for our application that is desired to run only in the main/browser process, not the renderer process. This PR will utilize a new fuse (https://github.com/electron/fuses/pull/8) called `LoadV8SnapshotFromCustomPath` that tells the main/browser process to look for the v8 snapshot file at `custom_v8_context_snapshot.bin`. Any other process will use the same path as is used today.

Consumers will then be able to build/package their binary as follows:

```js
flipFuses(
  path.join(options.dist, 'Cypress.app', 'Contents', 'MacOS', 'Cypress'), // Returns the path to the electron binary
  {
    version: FuseVersion.V1,
    [FuseV1Options.LoadBrowserProcessSpecificV8Snapshot]: true,
  },
)
```

Fixes #35170

#### Checklist

- [X] PR description included and stakeholders cc'd
- [ ] `npm test` passes
- [ ] tests are [changed or added](https://github.com/electron/electron/blob/main/docs/development/testing.md)
- [ ] relevant documentation is changed or added
- [X] [PR release notes](https://github.com/electron/clerk/blob/master/README.md) describe the change in a way relevant to app developers, and are [capitalized, punctuated, and past tense](https://github.com/electron/clerk/blob/master/README.md#examples).

#### Release Notes

Notes: Added `LoadBrowserProcessSpecificV8Snapshot` as a new fuse that will let the main/browser process load its v8 snapshot from a file at `browser_v8_context_snapshot.bin`. Any other process will use the same path as is used today.

### CODE REVIEW

**Comment by ryanthemanuel** [comment]
Note for reviewers: Currently, this just works on Macs, but provides an idea of what I'm trying to accomplish. It does not work on Linux because the renderer process is created from a zygote process that presumably is created prior to when the BrowserWindow arguments are specified.

There are a couple of questions I have for this PR before it is ready to go:

1. What approach would be recommended to handle linux? I have a version of code locally that works but it involves disabling the zygote process when the `custom-v8-snapshot-file-name` argument is present but I'm wondering if there's a better way.
2. Would it be better to make this a first class parameter under `webPreferences` rather than rely on `additionalArguments`. This may affect any docs that might be written so I haven't written up any docs just yet.
3. Being new to the code base, I'm curious if there's a good place for a test for this.

**Comment by codebytere** [comment]
@ryanthemanuel could you speak a little more to your use case here? what would you be trying to accomplish with per-process snapshots?

**Comment by ryanthemanuel** [comment]
@codebytere, we are using a custom snapshot in our main process to preload node dependencies (similar to what atom does with [electron-link](https://github.com/atom/electron-link)). However, we have `nodeIntegration` disabled in our renderer process and do *not* want those preloaded dependencies to be available in that context. Thus, we would either want the default snapshot (or no v8 snapshot) to be used in the renderer process.

**Comment by brian-mann** [comment]
@codebytere does Ryan’s update provide you with more context regarding our use case?

the original issue we opened with more description and context is here: https://github.com/electron/electron/issues/35170

**Comment by MarshallOfSound** [comment]
Based on the use case you described I don't think this PR is the right way forward.  Allowing custom snapshots via a command line flag in the renderer process is a can of worms and will not function correctly (As I think you've already identified in a few scenarios)

* Sandboxing
* Site isolation
* Delayed renderer creation
* Etc.

`additionalArguments` is a flawed property that I'm surprised survived the renderer-process-reuse refactor 🤔 

I think the correct approach here is to introduce a new `fuse` that specifically can change which snapshot is loaded in the **browser** process.  The renderer must always use the default snapshot (or at least a modified version of that snapshot) because it contains core blink/v8 builtins.  Adding a fuse that when enabled loads a _different_ v8 snapshot in the browser process seems like a minimal lift and slightly more explainable than a command line argument to a web contents (which fundamentally doesn't align with our modern process model)

**Comment by ryanthemanuel** [comment]
@MarshallOfSound, the `fuse` was exactly what I was looking for. I originally had explored going down the route of the custom v8 snapshot path being for the browser process, but I couldn't figure out how to make it work without passing in a parameter on launch. I hadn't heard of `fuses` before but a new one seems like it will solve that exact problem. 

I tried it out on mac and linux and it works. The code I used is now attached to this PR; however, I think there are still some discussion points with it before I think it's fully ready for review:

* Right now I'm calling the `fuse` `LoadV8SnapshotFromCustomPath`. I was thinking maybe `LoadBrowserProcessV8SnapshotFromCustomPath` to be even more explicit. Thoughts?
* The custom path for the v8 snapshot will be at `custom_v8_context_snapshot.bin` not including any of the architecture info in the filename. Does that seem reasonable?
* Do we still want to communicate the custom path via a [command line parameter](https://github.com/electron/electron/pull/35266/files#diff-9d44bd6ccb0600ec1500a16feb30b768aefdb679498646a47e7794d5e024c5dcR30-R31)? I have it working that way currently, but wasn't sure if there was a better option. Maybe using the `electron_main_delegate` or `ContentMainParams` as a method to communicate the path?

**Comment by ryanthemanuel** [comment]
@MarshallOfSound I went ahead and refactored to use the delegate instead of a command line parameter to communicate the custom v8 snapshot path. It seemed to simplify things fairly well. Does this seem like a reasonable approach?

**Comment by ryanthemanuel** [comment]
Also, I created the corresponding PR to the [fuses repo](https://github.com/electron/fuses/pull/8)

**Comment by ryanthemanuel** [comment]
Hi @MarshallOfSound and @codebytere, is there any update on what needs to happen to get this PR moving? Is there any additional information you need from me? Does my current approach seem reasonable?

**Comment by brian-mann** [comment]
@MarshallOfSound and @codebytere, any update here?

File: shell/app/electron_main_delegate.cc [hunk 1]
```diff
@@ -422,6 +423,20 @@ absl::optional<int> ElectronMainDelegate::PreBrowserMain() {
...
+base::StringPiece ElectronMainDelegate::GetCustomV8SnapshotFilename() {
+  const base::CommandLine* command_line =
+      base::CommandLine::ForCurrentProcess();
+  std::string process_type =
+      command_line->GetSwitchValueASCII(::switches::kProcessType);
+  bool use_custom_v8_snapshot_file_name =
+      process_type.empty() &&
+      electron::fuses::IsLoadV8SnapshotFromCustomPathEnabled();
+  if (use_custom_v8_snapshot_file_name) {
+    return "custom_v8_context_snapshot.bin";
```
Comment (MarshallOfSound) [inline shell/app/electron_main_delegate.cc:435]: Can we call this `browser_v8_context_snapshot.bin`

File: build/fuses/fuses.json5 [hunk 2]
```diff
@@ -7,5 +7,6 @@
   "node_options": "1",
   "node_cli_inspect": "1",
   "embedded_asar_integrity_validation": "0",
-  "only_load_app_from_asar": "0"
+  "only_load_app_from_asar": "0",
+  "load_v8_snapshot_from_custom_path": "0"
```
Comment (MarshallOfSound) [inline build/fuses/fuses.json5:11]:
```suggestion
  "load_browser_process_specific_v8_snapshot": "0"
```

I think that name is better

**Review by MarshallOfSound — COMMENTED** [review]
Can you rebase and add documentation for this fuse in the docs folder

**Comment by ryanthemanuel** [comment]
@MarshallOfSound I made the changes you requested. Let me know if there's anything else I need to do. Thanks!

**Comment by ryanthemanuel** [comment]
(I also updated the corresponding PR here: https://github.com/electron/fuses/pull/8/files)

**Comment by ryanthemanuel** [comment]
Also, I was curious what the policy was for back patching something like this. Would it be possible to be back patched to v19? Or only v20?

**Comment by ckerr** [comment]
> Also, I was curious what the policy was for back patching something like this. Would it be possible to be back patched to v19? Or only v20?

The Releases Working Group considers each backport request separately & it's a team decision. The considerations are about what you'd expect -- smaller PRs are better than bigger ones, opt-in is better than enabled by default, fewer upstream patches are better than more upstream patches, etc. Basically it all boils down to risk.

My *guess* is that if this were merged into `main` for 21, and if you requested it to be backported to 20, that the WG would at least consider it for 20 but probably not for 19.

That guess is just to give some transparency on how the decisions are made, though. It's not possible to give a definitive answer here since it's a team decision.

**Review by MarshallOfSound — APPROVED** [review]

**Comment by ryanthemanuel** [comment]
@MarshallOfSound and @ckerr does the `no-backport` mean that this change will make it to 21.0 or 22.0? We'd really like to consume this as soon as we can and I know 22.0 won't land for another 10 weeks.

**Review by ckerr — APPROVED** [review]

**Review by codebytere — APPROVED** [review]

**Comment by codebytere** [comment]
API LGTM

**Review by ckerr — APPROVED** [review]
API LGTM

**Comment by ckerr** [comment]
> @MarshallOfSound and @ckerr does the no-backport mean that this change will make it to 21.0 or 22.0? We'd really like to consume this as soon as we can and I know 22.0 won't land for another 10 weeks.

I think it was to make the bot happy :slightly_smiling_face:

**Pull request merged**